### ISSUE TITLE
[VM] Incorrect type of ExecResult.exceptionError

### ISSUE THREAD

**Issue by alcuadrado** [ethereumjs/ethereumjs-monorepo#1157]
I mentioned this bug some time ago, and never got to the root cause of it, so I'm opening this issue describing it.

`ExecResult` has this field `exceptionError?: VmError`. This is incorrect, as sometimes the error has a different type. This was breaking Hardhat in very weird ways.

A way to reproduce it is to use a custom `StateManager` that throws on, say, `putContractStorage`. Somehow this is watched and ends up there, and the types lie to you.

The interpreter has this TODO btw `// TODO: Throw on non-VmError exceptions`. I assume it's the same thing.

**Comment by alcuadrado** [comment]
Also, IMO changing that type is a breaking change. Resolving that TODO and throwing on non-`VmError`s is probably also one and may break quite a few things.

**Comment by holgerd77** [comment]
Hi there,
thanks for reporting. Had a first look: a first obvious place for this to happen is [here](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/vm/lib/evm/evm.ts#L319) in the EVM class where we assign this eventual `errorMessage` object which is `any` and therefore not guarded by TypeScript.

**Comment by holgerd77** [comment]
Update: ah, you are right, this can only result from within the `Interpreter` where the `exceptionError` is taken over from the `Interpreter` [here](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/vm/lib/evm/evm.ts#L450) within the `EVM` class.

Mentioned TODO can be found [here](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/vm/lib/evm/interpreter.ts#L122) in the `Interpreter` code.

**Comment by holgerd77** [comment]
I followed the call chain and what errors are guarded where by `try / catch`, is there any reason to not just throw along the mentioned code from the TODO respectively check for a `VMError` and throw on all other cases? 🤔

```typescript
let err
// Iterate through the given ops until something breaks or we hit STOP
while (this._runState.programCounter < this._runState.code.length) {
  const opCode = this._runState.code[this._runState.programCounter]
  this._runState.opCode = opCode
  await this._runStepHook()

  try {
    await this.runStep()
  } catch (e) {
    // STOP is not an exception
    if (e.error !== ERROR.STOP) {
      err = e
    }
    // TODO: Throw on non-VmError exceptions
    break
  }
}

return {
  runState: this._runState,
  exceptionError: err,
}
```

So add an `if error !== VM error throw` (pseudo-code 😄 ) where the TODO stands?

If an error is a non-VMError something went clearly wrong I guess, and the only thing we are doing here is guarding false VM behavior? 🤔

**Comment by holgerd77** [comment]
Have triggered a test on this along #1168 to see what will happen on CI run.

**Comment by holgerd77** [comment]
Closed by #1168
